src/tvp/model.ts

```diff
@@ -0,0 +1,352 @@
+import { TextDocument, Position } from 'vscode-languageserver-types'
+
+import { TreeViewProvider } from './provider'
+import { TreeViewNode } from './domain'
+import { groupBy } from './utils'
+
+export interface NodeView {
+  underlying: TreeViewNode
+  level: number
+  expandable: boolean
+  expanded: boolean
+}
+
+export interface TreeModelUpdate {
+  root: NodeView
+  oldNodes: NodeView[]
+  newNodes: NodeView[]
+  focusEvent: boolean
+}
+
+export class Node {
+  private expanded: boolean
+  private children: Promise<Node[]> | undefined
+  private lastTouch = Date.now()
+  private defunct = false
+
+  constructor(
+    public viewNode: TreeViewNode,
+    public level: number,
+    private provider: TreeViewProvider,
+    private emitter: Emitter<TreeModelUpdate>
+  ) {
+    this.expanded = false
+  }
+
+  public expandable(): boolean {
+    return this.viewNode.collapseState !== undefined
+  }
+
+  public isExpanded(): boolean {
+    return this.expanded
+  }
+
+  public async expand(): Promise<boolean> {
+    if (this.defunct) return false
+
+    this.provider.sendTreeNodeVisibilityNotification(this.viewNode.nodeUri, false)
+    this.expanded = true
+    const curState = await this.touch()
+    await this.getChildren()
+    const newNodes = await this.collectVisibleNodes()
+    if (curState === this.lastTouch && !this.defunct) {
+      this.emitter.fire({root: this.makeView(), oldNodes: [this.makeView()], newNodes, focusEvent: false})
+      return true
+    } else {
+      return false
+    }
+  }
+
+  public async collapse(): Promise<boolean> {
+    if (this.defunct) return false
+
+    this.provider.sendTreeNodeVisibilityNotification(this.viewNode.nodeUri, true)
+    const curState = await this.touch()
+    const oldNodes = await this.collectVisibleNodes()
+    this.expanded = false
+    if (curState === this.lastTouch && !this.defunct) {
+      this.emitter.fire({root: this.makeView(), oldNodes, newNodes: [this.makeView()], focusEvent: false})
+      return true
+    } else {
+      return false
+    }
+  }
+
+  public async refreshSubtree(viewNode: TreeViewNode | undefined): Promise<void> {
+    async function findOpenNodes(acc: string[][], curPath: string[], node: Node): Promise<string[][]> {
+      if (node.isExpanded()) {
+        acc.push(curPath)
+        const children = await node.getChildren()
+        return children.reduce((zP, child) => {
+          return zP.then(z => findOpenNodes(z, curPath.concat(child.viewNode.nodeUri), child))
+        }, Promise.resolve(acc))
+      } else {
+        return acc
+      }
+    }
+
+    async function reloadSubtree(node: Node, openNodes: string[][]): Promise<void> {
+      if (openNodes.length != 0) {
+        const children = await node.getChildren()
+        const childToNodes = groupBy(openNodes.filter(node => node.length != 0), arr => arr[0])
+        await Promise.all(children.map(child => {
+          const mbOpenNodes = childToNodes.get(child.viewNode.nodeUri)
```

Comment:
Same comment as above. I think for future maintainability, changing mb will be easier for a ready to quickly skim and see what it's referring to.

Reply:
`mb` means maybe. Type of this variable `string | undefined` and this name helps to remember that we should check variable for `null`. There is even special option in typescript compiler which requires to check null-reference explicitly but this option is not active. So this name at least helps to remember about possible `null` value.

Reply:
Alright. Sounds good. Thanks for the explanation.